homeassistant/components/group/climate.py

```diff
@@ -0,0 +1,396 @@
+        ]
+        # return the most common hvac mode (what the thermostat is set to do) except OFF, UNKNOWN and UNAVAILABE
+        if current_hvac_modes:
+            self._attr_hvac_mode = HVACMode(
+                max(sorted(set(current_hvac_modes)), key=current_hvac_modes.count)
+            )
+        # return off if any is off
+        elif any(x.state == HVACMode.OFF for x in states):
+            self._attr_hvac_mode = HVACMode.OFF
+        # else it's none
+        else:
+            self._attr_hvac_mode = None
+        # return the most common action if it is not off
+        hvac_actions = list(find_state_attributes(states, ATTR_HVAC_ACTION))
+        current_hvac_actions = [a for a in hvac_actions if a != HVACAction.OFF]
+        # return the most common action if it is not off
+        if current_hvac_actions:
+            self._attr_hvac_action = max(
+                sorted(set(current_hvac_actions)), key=current_hvac_actions.count
+            )
+        # return action off if all are off
+        elif all(a == HVACAction.OFF for a in hvac_actions):
+            self._attr_hvac_action = HVACAction.OFF
+        # else it's none
+        else:
+            self._attr_hvac_action = None
+
+        # available swing modes
+        all_swing_modes = list(find_state_attributes(states, ATTR_SWING_MODES))
+        if all_swing_modes:
+            self._attr_swing_modes = merge_modes(all_swing_modes)
+
+        # Report the most common swing_mode.
+        self._attr_swing_mode = most_frequent_attribute(states, ATTR_SWING_MODE)
+
+        # available fan modes
+        all_fan_modes = list(find_state_attributes(states, ATTR_FAN_MODES))
+        if all_fan_modes:
+            self._attr_fan_modes = merge_modes(all_fan_modes)
+
+        # Report the most common fan_mode.
+        self._attr_fan_mode = most_frequent_attribute(states, ATTR_FAN_MODE)
+
+        # available preset modes
+        all_preset_modes = list(find_state_attributes(states, ATTR_PRESET_MODES))
+        if all_preset_modes:
+            self._attr_preset_modes = merge_modes(all_preset_modes)
+
+        # Report the most common fan_mode.
+        self._attr_preset_mode = most_frequent_attribute(states, ATTR_PRESET_MODE)
+
+        # Supported flags
+        for support in find_state_attributes(states, ATTR_SUPPORTED_FEATURES):
+            # Merge supported features by emulating support for every feature
+            # we find.
+            self._attr_supported_features |= support
+
+        # Bitwise-and the supported features with the Grouped climate's features
+        # so that we don't break in the future when a new feature is added.
+        supported_features = ClimateEntityFeature(0)
+        for feature_flags, entity_set in self._features.items():
+            if entity_set:
+                supported_features |= feature_flags
+        self._attr_supported_features = supported_features
+
+    async def async_set_temperature(self, **kwargs: Any) -> None:
```

Comment:
I wonder a bit about this one as we can send temp-range to entities which doesn't support it.
Should we not be more explicit about only sending what is actually supported?

Reply:
Given that we are at the end calling the same method/service for all entities with all data (target value and range), each entity implementation would consume the values it needs ([example](https://github.com/home-assistant/core/blob/dev/homeassistant/components/ecobee/climate.py#L703-L707)), we can have here a simple logic that would call the same service with all the available data to all the entities in the group

Reply:
I don't think you entirely understand my point.
Looking at `sensibo` as example it doesn't support `target_range` and therefore it would raise if you call it with range values, see https://github.com/home-assistant/core/blob/dev/homeassistant/components/sensibo/climate.py#L321

I don't think that's very user friendly

Reply:
Agreed that would not be very user friendly, but I would argue that here we are talking about a more fundamental API problem. Having a method that contains a freeform of arguments being passed (through the `**kwargs`) without any guarantees about which parameters are being passed independently of the supported features of each entity type, leads to individual implementations handling their optionality in different ways.

With some [search](https://github.com/search?q=repo%3Ahome-assistant%2Fcore+symbol%3Aasync_set_temperature+path%3Aclimate&type=code&p=2), you can see that there are different implementations handling the presence and missing of parameters in different ways:

1. If not getting the value you need, do nothing: [code](https://github.com/home-assistant/core/blob/b0a1450a2b654d338d6fee7033b675b73275cb1a/homeassistant/components/generic_thermostat/climate.py#L395)
2. Silently raising a `KeyError` exception: [examples](https://github.com/search?q=repo%3Ahome-assistant%2Fcore+symbol%3Aasync_set_temperature+path%3Aclimate+%22kwargs%5BATTR_TEMPERATURE%5D%22&type=code)
3. Pass the target temperature value or `None` down the line to the underlying coordinator: [code](https://github.com/home-assistant/core/blob/f0b07ae942317aea31e24741d4c211ac065d3eaf/homeassistant/components/atag/climate.py#L86)
4. And others around how to handle the temperature range I did not went too far to dig in....

The point you just raise about the exception being raised would have also happened if we were to make a service call to a list of entity types and providing all parameters in the schema (at the end a group is doing that when its service is being called).  If we call a `sensibo` device with an `ecobee` with all parameters in the schema, we would get the methods called, but because [of this line](https://github.com/home-assistant/core/blob/f01235ef746823a8f146c603bf732df6a6061706/homeassistant/helpers/service.py#L996) we would not be calling the methods to update the state in the state machine, which again is not very user friendly.

My point (and preference) here is that in an ideal world, logic should be simple everywhere rather than complex. Specific climate platforms should not raise Exceptions if not supported parameters are being passed, as well as the group should just forward whatever parameters have been given to all the entities in the group (more future proof in case the schema of the call gets extended).

I can update the logic and make it more complex to take into account the supported features for each entity in the group so we make a targeted call for each entity, but this would not solve the bigger underlying problem that set temperature in a Climate entity ID already has and this change is not enlarging.

Let me know if I should proceed with updating the logic.

Reply:
Alternatively, the logic to decide which parameters should be passed to the `async_temperature_set` method given the supported features, could be placed [here](https://github.com/home-assistant/core/blob/f01235ef746823a8f146c603bf732df6a6061706/homeassistant/components/climate/__init__.py#L902-L917) which would take affect across the board for all Climate entities to do the right thing and avoid the mentioned problem. This most likely should be a separate PR.

Reply:
Proposed https://github.com/home-assistant/core/pull/114538 to handle this logic more natively on the ClimateEntity implementation.

Reply:
the schema handles that we can't use temperature and a temp range at the same time.
So my thinking was simply to check if it's a single temp or a temp range, then check the corresponding supported entities and send the call only to the entities which has support.

note that all these are optional from the schema so you need to something like
```python
if kwargs.get(ATTR_TEMPERATURE):
    data = {
            ATTR_ENTITY_ID: self._features[
                ClimateEntityFeature.TARGET_TEMPERATURE
            ]
        }
elif kwargs.get(ATTR_TARGET_TEMP_LOW):
    data = {
            ATTR_ENTITY_ID: self._features[
                ClimateEntityFeature.TARGET_TEMPERATURE_RANGE
            ]
        }
...
```

Reply:
Can we bring the discussion about what arguments to pass to the `set_temperature` method to the architectural proposal (https://github.com/home-assistant/architecture/discussions/1071) given that this logic seems more appropriate to be defined at the base ClimateEntity logic rather than on this children specialisation. The proposal would remove any need to handle this logic on this change.